File: Andy/src/level1/java/KeyPadPress.java
package level1.java;

public class KeyPadPress {
    public String solution(int[] numbers, String hand) {
		int[] left = {3,0};
		int[] right = {3,2};
		
        String answer = "";
		final String Left = "L";
		final String Right = "R";   
        final int PY = 0;
        final int PX = 1;
        if(hand.compareTo("left") == 0) hand = Left;
        else hand = Right;
		
        int[] loc = {0,0};
        String ans = Left;
        
        for(int d:numbers) {
            loc[PY] = d==0?3:(d-1)/3;
            loc[PX] = d==0?1:(d-1)%3;
            
            if(loc[PX] == 0) 
            	ans = Left;
            else if(loc[PX] == 2)
            	ans = Right;
            else {           
            	int left_len = Math.abs(loc[PX]-left[PX]) + Math.abs(loc[PY]-left[PY]); 
            	int right_len = Math.abs(loc[PX]-right[PX]) + Math.abs(loc[PY]-right[PY]);
                
            	if(left_len<right_len) 
            		ans = Left;
            	else if(left_len>right_len) 
            		ans = Right;
            	else
            		ans = hand;
            }
            
            answer += ans;
            if(ans == Left) 
                left = loc.clone();
            else 
                right = loc.clone();
        }
        
        return answer;
    }
}

File: Andy/src/level1/javascript/KeyPadPress.js
//키패드 누르기
// 좌표 : [ 행,열 ]
function solution(numbers, hand) {
    const Left = "L";
    const Right = "R";
    var answer = '';

    var finger=[];
    finger[Left] = [ 3, 0 ];
    finger[Right] = [ 3,2 ];
    finger.cal = function(arrow,loc) { 
        return Math.abs(loc[0]-this[arrow][0]) + Math.abs(loc[1]-this[arrow][1]);
    } 
  
    numbers.forEach( function(d) { 
        let ans = Left;    
        var loc = [ parseInt(d==0?3:(d-1)/3) , parseInt(d==0?1:(d-1)%3) ];
        
        if(loc[1] == 0) 
            ans = Left;
        else if(loc[1] == 2) 
            ans = Right;           
        else {
            let left_len = finger.cal(Left,loc);
            let right_len = finger.cal(Right,loc);
            if(left_len<right_len) 
                ans = Left;
            else if(left_len>right_len) 
                ans = Right;
            else 
                ans = hand=="right"?Right:Left;
        }
         
        answer += ans;
        finger[ans] = loc;
    });
    
    return answer;
}